File: lab7.py
class CircularQueue:
    def __init__(self, size):
        self.data = [None] * size
        self.size = size
        self.start = size // 2
        self.end = size // 2 + 1


    def enqueue(self, item):
        if self.data[self.start] == None:
            self.data[self.start] = item
        else:
            self.data[self.end] = item
            if self.end + 1 == self.size:
                self.end = 0
            else:
                self.end += 1

    def dequeue(self):
        # return self.data.pop()
        ret_val = self.data[self.start]
        self.data[self.start] = None # O(n), because need to move self.data[1:] to self.data[0:]
        if self.start + 1 >= self.size:
            self.start = 0
        else:
            self.start += 1
        return ret_val

    def __str__(self):
        # res = ""
        # res += "index\t"
        # for i in range(self.size):
        #     res += f"{i}\t"
        # res += "\n"
        # for i in range(self.size):
        #     if(self.data[i] == None):
        #         res += "\t"
        #     else:
        #         res += f"{self.data[i]}\t"
        # res += "\n"
        # res += f"Begin at index: {self.start}\n"
        # res += f"End at index: {self.end}\n"
        # return res
        if(self.end >= self.start):
            self_size = self.end - self.start
        else:
            self_size = self.end + self.size - self.start
        ind = self.start
        res = ""
        for i in range(self_size-1):
            res += str(self.data[ind]) + "->"
            if ind + 1 >= self.size:
                ind = 0
            else:
                ind += 1
        res += str(self.data[ind])
        return res


    def __lt__(self, other):
        ind_self = self.start
        ind_other = other.start
        
        while(ind_self != self.end+1 or ind_other != other.end+1):
            if self.data[ind_self] != other.data[ind_other]:
                return self.data[ind_self] < other.data[ind_other]
            else:
                ind_self += 1
                ind_other += 1
                if ind_self >= self.size:
                    ind_self = 0
                if ind_other >= other.size:
                    ind_other = 0
        return True

my_queue = CircularQueue(10)
my_queue.enqueue(6)
print(my_queue)
my_queue.enqueue(2)
print(my_queue)
my_queue.enqueue(7)
print(my_queue)
my_queue.dequeue()
print(my_queue)

File: lab7.c
# include <stdio.h>
# include <stdlib.h>
# include <string.h>

typedef struct CircularQueue{
    int *arr;
    int size;
    int start;
    int end;
} CircularQueue;

void init(CircularQueue *q, int size){
    q->arr = (int *)malloc(size * sizeof(int));
    for(int i = 0; i < size; i++){
        q->arr[i] = 0;
    }
    q->size = size;
    q->start = size / 2;
    q->end = size / 2 + 1;
}

void enqueue(CircularQueue *q, int item){
    if(q->arr[q->start] == 0){
        q->arr[q->start] = item;
    }
    else{
        q->arr[q->end] = item;
        if(q->end+1 == q->size){
            q->end = 0;
        }
        else{
            q->end++;
        }
    }
}

int dequeue(CircularQueue *q){
    int res = q->arr[q->start];
    q->arr[q->start] = 0;

    if(q->start+1 == q->size){
        q->start = 0;
    }
    else{
        q->start++;
    }
    return res;
}

void printQueue(CircularQueue *q){
    int q_size = 0;
    if(q->end >= q->start){
        q_size = q->end - q->start;
    }
    else{
        q_size = q->size - q->start + q->end;
    }
    int ind = q->start;
    for(int i = 0; i < q_size-1; i++){
        printf("%d->", q->arr[ind]);
        if(ind+1 == q->size){
            ind = 0;
        }
        else{
            ind++;
        }
    }
    printf("%d\n", q->arr[ind]);
}

int queuecmp(CircularQueue *q1, CircularQueue *q2){
    int q1_ind = q1->start;
    int q2_ind = q2->start;

    while(q1_ind != q1->end+1 || q2_ind != q2->end+1){
        if(q1->arr[q1_ind] < q2->arr[q2_ind]){
            return -1;
        }
        else if(q1->arr[q1_ind] > q2->arr[q2_ind]){
            return 1;
        }
        else{
            q1_ind++;
            q2_ind++;
            if(q1_ind >= q1->size){
                q1_ind = 0;
            }
            if(q2_ind >= q2->size){
                q2_ind = 0;
            }
        }
    }
    return 0;
}

void expand(CircularQueue *q){
    CircularQueue *new_q = (CircularQueue *)malloc(sizeof(CircularQueue));
    new_q->arr = (int *)malloc(q->size * 2 * sizeof(int));
    int q_size = 0;
    if(q->end >= q->start){
        q_size = q->end - q->start;
    }
    else{
        q_size = q->size - q->start + q->end;
    }

    int ind = q->start;
    for(int i = 0; i < q_size; i++){
        new_q->arr[i] = q->arr[ind];
        ind++;
        if(ind == q->size){
            ind = 0;
        }
    }
    new_q->size = q->size * 2;
    new_q->start = q->start;
    if(new_q->start + q_size >= new_q->size){
        new_q->end = q_size - (new_q->size - new_q->start);
    }
    else{
        new_q->end = new_q->start + q_size;
    }
    q = new_q;
}

int main(){
    CircularQueue q1, q2;
    init(&q1, 10);
    init(&q2, 10);

    enqueue(&q1, 1);
    enqueue(&q1, 2);
    enqueue(&q1, 3);
    enqueue(&q1, 4);
    enqueue(&q1, 5);


    enqueue(&q2, 6);
    enqueue(&q2, 7);
    enqueue(&q2, 8);
    enqueue(&q2, 9);
    enqueue(&q2, 10);

    printQueue(&q1);
    printQueue(&q2);

    printf("%d\n", queuecmp(&q1, &q2));

    return 0;
}